zstd: Update zstd to 1.5.6 for cmake 4 compatability

edwinzrodriguez opened the pull request:
cmake 4 drops support for cmake_minimum_required < 3.5 This update bumps zstd to support newer cmake version

Fixes: https://tracker.ceph.com/issues/73522





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [X] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [X] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [X] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [X] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)

You must only issue one Jenkins command per-comment. Jenkins does not understand
comments with more than one command.
</details>

cbodley reviewed (approved):
the rgw suite has coverage of all compressor plugins, we can use that to validate

tchaikov reviewed (commented):
i'd suggest split this change into two commits.

1. the first commit migrates from the undeprecated APIs. 
2. and the second one bumps up the submodule.

in the commit message of the first commit, we'd better explain why the API was deprecated by referencing the related release notes, and explain the new APIs are available by the minimal supported zstd version. 

also, would be better to

1. take this opportunity to use `ZSTD_createCCtx()` in favor of `ZSTD_createCStream()`.
2. use `ZSTD_CCtx_reset(cctx, ZSTD_reset_session_and_parameters)` instead of `ZSTD_CCtx_reset()` and `ZSTD_CCtx_refCDict()`
3. check the return value of these API calls.

edwinzrodriguez committed:
compressor/zstd: modernize ZSTD API usage in ZstdCompressor

Update ZstdCompressor to use the modern ZSTD compression context API
and improve error handling:

- Replace deprecated ZSTD_createCStream() with ZSTD_createCCtx()
- Use ZSTD_CCtx_reset() with ZSTD_reset_session_and_parameters flag
  instead of separate ZSTD_reset_session_only and ZSTD_CCtx_refCDict()
  calls. This consolidates session and parameter reset into a single
  operation.
- Add proper return value checking for all ZSTD API calls using
  ZSTD_isError() to catch compression failures
- Ensure proper cleanup with ZSTD_freeCCtx() on all error paths to
  prevent memory leaks
- Update corresponding free function from ZSTD_freeCStream() to
  ZSTD_freeCCtx()

These changes align with ZSTD's recommended API usage patterns and
improve robustness by properly handling potential compression errors.

Fixes: https://tracker.ceph.com/issues/73522
Signed-off-by: Edwin Rodriguez <edwin.rodriguez1@ibm.com>

edwinzrodriguez committed:
zstd: update submodule to v1.5.6

Update zstd submodule to version 1.5.6 for improved performance,
bug fixes, and modern API support.

Fixes: https://tracker.ceph.com/issues/73522
Signed-off-by: Edwin Rodriguez <edwin.rodriguez1@ibm.com>

tchaikov reviewed (approved):
lgtm with a nit. but it's not a blocker.

edwinzrodriguez commented:
jenkins test make check

tchaikov commented:
@edwinzrodriguez i don't think it's been picked up for integration tests after checking the comments and labels in this PR. why do you think it has been tested in lab?

JonBailey1993 commented:
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues73711
